### ISSUE TITLE
Incorrect displaying of out of stock bundle product 

### ISSUE THREAD

**Issue by staradzinau** [magento/magento2#33117]
### Preconditions (*)

1. Magento 2.4.2 with sample data installed
2. Stores → Configuration → Catalog → Inventory → Stock Options → Display Out of Stock Products → set the option to NO
3. Simple product, in stock, but with limited qty. Example: 
![image](https://user-images.githubusercontent.com/24728606/120326668-a4a33b80-c2f1-11eb-9dae-1f01672584e3.png)

### Steps to reproduce (*)

1. Create bundle product with simple product from the preconditions
2. For this bundle, make the product as a required option, forbid to modify qty by user, and set default qty > current stock qty for simple product 
![image](https://user-images.githubusercontent.com/24728606/120326777-c270a080-c2f1-11eb-818d-9100a01e7652.png)
3. Run reindex
4. Go to the storefront and open any category, where created bundle product is assigned to

### Expected result (*)

2. Bundle product should be invisible in the catalog, according to the config option "Display Out of Stock Products"

### Actual result (*)

1. Bundle product is visible in the catalog, but marked as "out of stock", while option "Display Out of Stock Products" is set to NO.
![image](https://user-images.githubusercontent.com/24728606/120327066-082d6900-c2f2-11eb-88e9-658dad0cb252.png)

### Additional information
As I can see, stock index for bundle product does not take into account default qty of simple product, specified within the bundle

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by engcom-Alfa** [comment]
Tried in 2.4-develop version today and it is reproducible. Below are the reproducible steps.

1. Set the Stores → Configuration → Catalog → Inventory → Stock Options → Display Out of Stock Products → set the option to NO
2. Created 2 simple products with quantity value 5 as shown in the below screenshots. 
![image](https://user-images.githubusercontent.com/51679138/120444677-77b06080-c3a5-11eb-90e8-6f397da96f27.png)

3. Created bundle product using the above created simple products with the settings as shown in below screen shot (Where default quantity set to 10)
![image](https://user-images.githubusercontent.com/51679138/120444767-91ea3e80-c3a5-11eb-856b-ff8411f650cc.png)

4. Able to see the bundle product in the front end and also able to add the product to checkout. 
![image](https://user-images.githubusercontent.com/51679138/120444949-c5c56400-c3a5-11eb-9c64-980ae4a12d63.png)

Expected: Step 4 should not have been executed.
Basically since the simple products quantity set 5 each, and the bundle default quantities set to 10, the bundle product should not have been displayed in the frontend catalog products list page.
Else, it should come out with a message saying product is out of stock, while checking out from the shopping cart.

5. But, proceeding with checkout, it the order is getting placed
![image](https://user-images.githubusercontent.com/51679138/120446231-f8bc2780-c3a6-11eb-9f00-63aca91756ba.png)

Hence, confirming the issue.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Alfa 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-42530` were created

**Issue Available**: @engcom-Alfa, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._
